### PR TITLE
Make the MatchingWords return the number of matching bytes

### LINKED ISSUE #138: Highlight behavior
Currently, and contrary to the old MeiliSearch, Milli highlights the whole word instead of highlighting only the part of the word related to the query.

Ex: If the document contains `cheval blanc` and the query is `chev`, Milli will highlight this way:`<em>cheval</em> blanc`.
The old MeiliSearch would highlight `<em>chev</em>al blanc`

We need to think about a way to handle the highlighting as the old MeiliSearch does, to stay iso.

### PR DESCRIPTION
**Pull request by LegendreM**
Make the MatchingWords return the number of matching bytes using a custom Levenshtein algorithm.

Fix #138

### CODE REVIEW

File: milli/src/search/matching_words.rs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+impl MatchingWords {
+    /// Lists all words which can be considered as a match for the query tree.
+    pub fn from_query_tree(tree: &Operation) -> Self {
+        let mut dfas: Vec<_> = fetch_queries(tree)
+            .into_iter()
+            .map(|(w, t, p)| (build_dfa(w, t, p), w.to_string(), t, p))
+            .collect();
+        dfas.sort_unstable_by_key(|(_dfa, query_word, _typo, _is_prefix)| Reverse(query_word.len()));
```
Comment (Kerollmops) [inline milli/src/search/matching_words.rs:27]: Could you add a comment on top of this line explaining the reason why we reverse sort them by length, please?

File: milli/src/search/matching_words.rs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+use levenshtein_automata::{DFA, Distance};
+
+use crate::search::query_tree::{Operation, Query};
+
+use super::build_dfa;
+
+type IsPrefix = bool;
+
+/// The query tree builder is the interface to build a query tree.
```
Comment (Kerollmops) [inline milli/src/search/matching_words.rs:14]: This comment is invalid, can you change it, please?

File: milli/src/search/matching_words.rs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+#[derive(Default)]
```
Comment (Kerollmops) [inline milli/src/search/matching_words.rs:15]:
```suggestion
#[derive(Clone, Default)]
```

File: milli/src/search/matching_words.rs [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+pub struct MatchingWords {
+    dfas: Vec<(DFA, String, u8, IsPrefix)>,
+}
+
+impl MatchingWords {
+    /// Lists all words which can be considered as a match for the query tree.
```
Comment (Kerollmops) [inline milli/src/search/matching_words.rs:21]: This comment is also invalid, can you change it, please?

File: milli/src/search/matching_words.rs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    pub fn from_query_tree(tree: &Operation) -> Self {
+        let mut dfas: Vec<_> = fetch_queries(tree)
+            .into_iter()
+            .map(|(w, t, p)| (build_dfa(w, t, p), w.to_string(), t, p))
+            .collect();
+        dfas.sort_unstable_by_key(|(_dfa, query_word, _typo, _is_prefix)| Reverse(query_word.len()));
+        Self { dfas }
+    }
+
+    /// Returns the number of matching bytes if the word matches.
```
Comment (Kerollmops) [inline milli/src/search/matching_words.rs:31]:
```suggestion
    /// Returns the number of matching bytes if the word matches one of the query words.
```

File: milli/src/search/matching_words.rs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+impl<T> IndexMut<(usize, usize)> for N2Array<T> {
+    #[inline]
+    fn index_mut(&mut self, (x, y): (usize, usize)) -> &mut T {
+        &mut self.buf[(x * self.y_size) + y]
+    }
+}
+
+fn prefix_damerau_levenshtein(source: &[u8], target: &[u8]) -> (u32, usize) {
```
Comment (Kerollmops) [inline milli/src/search/matching_words.rs:97]: Could you add a doc-comment on this function, please?

File: milli/src/search/query_tree.rs [hunk 7]
```diff
@@ -294,48 +294,6 @@ fn synonyms(ctx: &impl Context, word: &[&str]) -> heed::Result<Option<Vec<Operat
     }))
 }
 
-/// The query tree builder is the interface to build a query tree.
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:297]: Could you fix all the warnings in the imports introduced when you moved the `MatchingWords` struct away, please?

File: milli/src/search/query_tree.rs [hunk 8]
```diff
@@ -951,39 +909,6 @@ mod test {
         assert_eq!(expected, query_tree);
     }
 
-    #[test]
```
Comment (Kerollmops) [inline milli/src/search/query_tree.rs:954]: Same here, could you fix the warnings in the imports, please?

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: milli/src/search/matching_words.rs (hunk 3 above)
Comment (LegendreM) [reply milli/src/search/matching_words.rs:16]: `DFA` doesn't implement `Clone` :/

**Review by Kerollmops — APPROVED** [review]
Seems good to me!

bors merge

**Pull request merged**